### ISSUE TITLE
Using NULL as a C function pointer doesn't work

### ISSUE THREAD

**Issue by brianhuffman** [GaloisInc/crucible#7]
The following C program (compiled to LLVM bitcode) works just fine with Crucible. (I am testing with the `crucible-integration` branch of saw-script.)
```
#include <stddef.h>

typedef int func(int i);

int inc (int x) {
	return x + 1;
}

void apply (func *f, int *x) {
	*x = f(*x);
}

void test (int *x) {
	apply(inc, x);
}
```
However, if we add a test for `NULL` to the definition of `apply`
```
void apply (func *f, int *x) {
	if (f != NULL) { *x = f(*x); }
}
```
then we get an error from crucible:
```
at internal: arithmetic comparison on incompatible values Typed {typedType = PtrTo (FunTy (PrimType (Integer 32)) [PrimType (Integer 32)] False), typedValue = ValIdent (Ident "5")} ValNull
```

Similarly, if we modify function `test` to pass a `NULL` argument to `apply`
```
void test (int *x) {
	apply(NULL, x);
}
```
then we get a different error:
```
at internal: argument type mismatch in call to ValSymbol (Symbol "apply") [Typed {typedType = PtrTo (FunTy (PrimType (Integer 32)) [PrimType (Integer 32)] False), typedValue = ValNull},Typed {typedType = PtrTo (PrimType (Integer 32)), typedValue = ValIdent (Ident "3")}] PtrTo (FunTy (PrimType Void) [PtrTo (FunTy (PrimType (Integer 32)) [PrimType (Integer 32)] False),PtrTo (PrimType (Integer 32))] False)
```

I also tried modifying `src/Lang/Crucible/LLVM/Translation.hs` to print out `argTypes` and `argTypes'` in the "argument type mismatch" error message. Here they are:
```
[RecursiveRepr LLVM_pointer, RecursiveRepr LLVM_pointer]
[FunctionHandleRepr [BVRepr 32] (BVRepr 32), RecursiveRepr LLVM_pointer]
```

**Comment by robdockins** [comment]
The core problem here is that function pointers are currently treated as incompatible with data pointers during the LLVM translation phase.  Function pointers are mapped directly into crucible values of `FunctionHandle` type, whereas data pointers are mapped into the `LLVM_pointer` type.  This is incorrect for a number of reasons, including the fact that it makes function pointers incomparable with data pointers (including NULL).  I believe it will also fail if one attempts to cast a function pointer to `void *` type, which should be allowed (I think?)

I believe the proper fix is to make sure that function pointers are translated into `LLVM_pointer` type, just as data pointers are.  To associate the callable Crucible function handle with the LLVM pointer, we should instead use a side lookup table, similar to the way arbitrary Haskell data can be associated with pointers using `Lang.Crucible.LLVM.MemModel.doMallocHandle`.

Semantic questions to make sure we get right:
  - What is the semantics of reading/writing through a function pointer?  I strongly suspect it is undefined.
  - What is allowed in terms of pointer arithmetic on function pointers?
  - What is the semantics of freeing a function pointer?  Also suspect it is undefined.
  - Does C/LLVM guarantee that code pointer types and data pointer types are compatible (i.e., can I cast pointer types to `void *` and back)? I think so, but am not sure.

**Comment by brianhuffman** [comment]
Fixed in f29e17a31af289f75f6eb9933b6e735a855171da.
